internal/data/metric.go

```diff
@@ -0,0 +1,637 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package data
+
+import (
+	otlpmetric "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+)
+
+// This file defines in-memory data structures to represent metrics.
+// For the proto representation see https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/metrics/v1/metrics.proto
+
+// MetricData is the top-level struct that is propagated through the metrics pipeline.
+// This is the newer version of consumerdata.MetricsData, but uses more efficient
+// in-memory representation.
+type MetricData struct {
+	resourceMetrics []*ResourceMetrics
+}
+
+func NewMetricData(resourceMetrics []*ResourceMetrics) MetricData {
+	return MetricData{resourceMetrics}
+}
+
+// MetricCount calculates the total number of metrics.
+func (md MetricData) MetricCount() int {
```

Comment:
This is for the internal format but the lowest common denominator for various metric formats is actually a data-point, we probably should have a similar helper for that too. That said it can wait until we really need it.

Reply:
Will add that if needed.